### ISSUE TITLE
Juvix loading proces

### ISSUE THREAD

**Issue by mariari** [heliaxdev/juvix#740]
** Current Process

Currently Juvix takes a frontend ML expression, translates it into an s-expression, desugares it, throws it into a context, before turning it into Core.

** Proposal 

I propose the following structure of our langauge

1. Read a recursive block
2. Convert the block into S-expressions
     - Note the original form either here or in the next step, as we will need it for proof
3. Minor Type Checking for Modules, along with Recurisve Group Checking
    - The Point of this is to properly setup for the next phase, along with having a correct feed order
4. Resolve opens to their Qualified State
     - So symbols which rely on open behavior get qualified away. 
     - Opens within a module structure are noted, 
         - Thus for top level opens, we immediately notify the image of a new open and prepare that.
         - For opens within some kind of generated module, we try to register it with the structure. Details of this are not clear to me yet.
     - For functions, we note the open, but resolve the symbols to their qualified state.     
4. Run Macro Expansion
    - This includes all macros such as infix resolution.
    - We may want a staging system where macros can compose better, however that will require further research.
5.  End At Base Juvix
     - At this point we have the frontend form fully desugared. This means that **non type level macros** are resolved
6. Fully Register the new forms in the recursive block
    - We would likely allocate stubs in the image saying what will come in, along with their original form before this step.
    - We can now replace this stub with their actual definition
    - This step can be exchanged with the step below it depending on how we want to do deal with things 
7. Run Frontend Translation Into Core term
    - This in the Haskell will be a normal Haskell datatype, however in the LLVM image, this too will be an s-expression. 
        - The reasoning for this is that Haskell's type system isn't strong enough to constrain an s-expression, but it can for its own sum types.
        - However for the LLVM version of the image we will compile this will be an s-expression for ease of serialization, along with any custom tooling we make there will likely be more consistent and with the power of the Juvix programming language. 
8. Run Type Checking
9. Run Optimization passes
    - This is determined by looking in the image, seeing what the global variables say about optimizing each piece along with any module overrides or local function overrides any piece of code may contain.  
10. Regiser Generated Code into a misc slot of the image for fast evaluation.

The Above passes should be good to cover our bases for any kind of **declaration**. However for loose statements or function calls, we need to skip the registration step and eval it on the spot. This may lead to declarations that we need to then deal with accordingly.

**Comment by cwgoes** [comment]
Have you considered how much we'll need to have in scope at a time for certain kinds of optimisations (e.g. a lot of inlining) to work? I think one recursive block as the unit which moves through this whole pipeline (if I understand your proposal correctly) may be far too little.

**Comment by mariari** [comment]
> Have you considered how much we'll need to have in scope at a time for certain kinds of optimisations (e.g. a lot of inlining) to work? I think one recursive block as the unit which moves through this whole pipeline (if I understand your proposal correctly) may be far too little.

We have an entire image around, so the recursive block can do inlining at stage 10. So all the code we have sent down before will still exists and we can inline away, depending on the settings.

**Comment by andy-morris** [comment]
It sounds like step 4 is going to replace all names with their fully-qualified variants. Am I reading that right?

Either way this all sounds good to me.

**Comment by mariari** [comment]
> It sounds like step 4 is going to replace all names with their fully-qualified variants. Am I reading that right?
> 
> Either way this all sounds good to me.

No, we are only fully qualifying names which aren't apart of the current module, so any opens get fully qualified, etc.

**Comment by cwgoes** [comment]
Part of the idealised pipeline discussions.
